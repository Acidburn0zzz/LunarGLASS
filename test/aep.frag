#version 310 es

precision highp float;

#extension GL_ANDROID_extension_pack_es31a : enable

// - GL_KHR_blend_equation_advanced
// + GL_OES_sample_variables
// - GL_OES_shader_multisample_interpolation
// + GL_EXT_geometry_shader
// + GL_EXT_shader_io_blocks

in inName {
    vec4 color;
} inInst;

out vec4 color;

int foo_GS();
void goodSample();
vec3 interp();

void main()
{
    color = inInst.color;
    color *= float(foo_GS());
    goodSample();
    color.xyz += interp();
}

// GL_OES_geometry_shader

highp int foo_GS()
{
    highp int l = gl_Layer;
    highp int p = gl_PrimitiveID;
    return l + p;
}


// GL_OES_sample_variables

void goodSample()
{
    lowp     int  a1 = gl_SampleID;
    mediump  vec2 a2 = gl_SamplePosition; 
    highp    int  a3 = gl_SampleMaskIn[0];
    gl_SampleMask[0] = a3;
    mediump int n1 = gl_MaxSamples;
    mediump int n2 = gl_NumSamples;
    gl_SampleMask[0] += n1 + n2 + a1;
    color.xy += a2;
}

// GL_OES_shader_multisample_interpolation

#extension GL_OES_shader_multisample_interpolation : enable

in float scalarIn;

sample in vec4 colorSampIn;
flat sample in vec4 colorfsi;
sample in vec3 sampInArray;  // ??: add the array: "[4]", currently makes two "sample" come out
uniform int i;

vec3 interp()
{
    float res;
    vec3 res3;

    res   = interpolateAtCentroid(scalarIn);
    res3  = interpolateAtCentroid(sampInArray/*[2]*/);

    res3 += interpolateAtSample(sampInArray/*[i]*/, 0);
    res  += interpolateAtSample(scalarIn, 1);

    res3 += interpolateAtOffset(sampInArray/*[2]*/, vec2(0.2));
    res  += interpolateAtOffset(scalarIn, vec2(0.2));

    return res * res3;
}

layout(blend_support_darken) out;
layout(blend_support_all_equations) out;
