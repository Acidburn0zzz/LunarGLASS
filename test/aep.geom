#version 310 es

// + GL_EXT_geometry_shader
// + GL_EXT_geometry_point_size

#extension GL_ANDROID_extension_pack_es31a : enable

precision mediump float;

in fromVertex {
    in vec3 color;
} fromV[];

in vec4 nonBlockUnsized[];

out toFragment {
    out vec3 color;
} toF;

out fromVertex {
    centroid vec3 color;
    int len;
    int inv;
};

layout(triangle_strip) out;
layout(max_vertices = 200) out;
layout(lines_adjacency) in;
layout(invocations = 4) in;

centroid in vec3 centr[];

void pointSize2();

void main()
{
    EmitVertex();
    EndPrimitive();

    color = fromV[0].color + centr[1];
    gl_Position = gl_in[0].gl_Position;

    gl_PrimitiveID = gl_PrimitiveIDIn;
    gl_Layer = 2;

    len = fromV.length();     // 4: lines_adjacency
    inv = gl_InvocationID;

    pointSize2();
}


#extension GL_OES_geometry_point_size : enable

void pointSize2()
{
    highp float ps = gl_in[3].gl_PointSize;
    gl_PointSize = ps;
}
