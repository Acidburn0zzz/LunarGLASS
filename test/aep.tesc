#version 310 es

#extension GL_ANDROID_extension_pack_es31a : enable

// GL_EXT_tessellation_shader: patch
// GL_EXT_tessellation_point_size
// GL_OES_primitive_bounding_box

layout(vertices = 4) out;
int outa[gl_out.length()];

patch out vec4 patchOut;

void bb();
void pointSize();
float foop();

void main()
{
    barrier();

    vec4 p = gl_in[1].gl_Position;

    int pvi = gl_PatchVerticesIn;
    int pid = gl_PrimitiveID;
    int iid = gl_InvocationID;

    gl_out[1].gl_Position = p;

    gl_TessLevelOuter[3] = 3.2;
    gl_TessLevelInner[1] = 1.3;
    bb();
    pointSize();
    patchOut = vec4(0.5);
    patchOut *= foop();
}

in vec2 inb[];
in vec2 ind[gl_MaxPatchVertices];

#extension GL_ARB_separate_shader_objects : enable

layout(location = 3) in vec4 ivla[];
layout(location = 4) in vec4 ivlb[];

layout(location = 3) out vec4 ovla[];
layout(location = 4) out vec4 ovlb[];

patch out pinbn {
    int a;
} pinbi;

centroid out vec3 myColor2[];
centroid in vec3 centr[];
out float okaySize[4];

#extension GL_OES_tessellation_point_size : enable

// GL_EXT_tessellation_point_size

void pointSize()
{
    gl_out[1].gl_PointSize = gl_in[1].gl_PointSize;
}

// for testing with gpu_shader5
in vec3 inv[];
precise vec3 pv3;

float foop()
{
    precise float d;

    pv3 = inv[3];
    pv3 *= pv3;
    pv3 = fma(pv3, pv3, pv3);
    d = fma(pv3.x, pv3.y, pv3.z);

    return d;
}

// GL_OES_primitive_bounding_box

void bb()
{
    gl_BoundingBoxOES[0] = vec4(0.0);
    gl_BoundingBoxOES[1] = vec4(1.0);
}
