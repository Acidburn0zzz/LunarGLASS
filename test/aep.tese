#version 310 es

#extension GL_ANDROID_extension_pack_es31a : enable

// - GL_EXT_tessellation_shader
// + GL_EXT_tessellation_point_size

layout(quads, ccw) in;  // - no ccw

layout(fractional_odd_spacing) in;    

layout(point_mode) in;

patch in vec4 patchIn;

void pointSize();

void main()
{
    vec4 p = gl_in[1].gl_Position;

    int pvi = gl_PatchVerticesIn;
    int pid = gl_PrimitiveID;
    vec3 tc = gl_TessCoord;
    float tlo = gl_TessLevelOuter[3];
    float tli = gl_TessLevelInner[1];

    gl_Position = p;
    pointSize();
}

in vec2 inb[];
in vec2 ind[gl_MaxPatchVertices];

in testblb {
    int f;
} blb[];

in testbld {
    int f;
} bld[gl_MaxPatchVertices];

layout(location = 23) in vec4 ivla[];
layout(location = 24) in vec4 ivlb[];

layout(location = 23) out vec4 ovla[2];

patch in pinbn {
    int a;
} pinbi;

centroid out vec3 myColor2;
centroid in vec3 centr[];

#extension GL_OES_tessellation_point_size : enable

void pointSize()
{
    gl_PointSize = gl_in[1].gl_PointSize;
}
