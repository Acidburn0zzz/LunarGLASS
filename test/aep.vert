#version 310 es

#extension GL_ANDROID_extension_pack_es31a : enable

// GL_OES_shader_image_atomic
// GL_OES_texture_storage_multisample_2d_array
// GL_EXT_gpu_shader5
// GL_EXT_shader_io_blocks
// GL_EXT_texture_buffer
// GL_EXT_texture_cube_map_array

uniform sampler2D sArray[4];
uniform ubName { vec2 p; } ubInst[4];
uniform writeonly lowp image2D iArray[5];
uniform int index;

const ivec2 constOffsets[4] = ivec2[4](ivec2(1), ivec2(2), ivec2(3), ivec2(4));

in vec2 inf, ing, inch;

out outName {
    vec4 color;
} outInst;

vec4 bufferT();
vec4 CAT();
vec4 MSA();
void goodImageAtom();

void main()
{
    // GL_EXT_gpu_shader5
    precise vec2 p = fma(inf, ing, inch);
    iArray[index];
    vec4 color = vec4(0.0);
    color += textureGatherOffset(sArray[index], ubInst[index].p, ivec2(inf));
    color += textureGatherOffsets(sArray[index], p, constOffsets);
    outInst.color = color;
    outInst.color += bufferT();
    outInst.color += CAT();
    outInst.color += MSA();
    goodImageAtom();
}

// GL_EXT_texture_buffer

uniform highp samplerBuffer  bufSamp1;          
uniform highp isamplerBuffer bufSamp2;          
uniform highp usamplerBuffer bufSamp3;          
uniform highp writeonly imageBuffer    bufSamp4;
uniform highp writeonly iimageBuffer   bufSamp5;
uniform highp writeonly uimageBuffer   bufSamp6;

vec4 bufferT()
{
    vec4 v = vec4(1.0);

    highp int s1;
    s1 = textureSize(bufSamp1);
    v *= vec4(s1);
    s1 = textureSize(bufSamp2);
    v *= vec4(s1);
    s1 = textureSize(bufSamp3);
    v *= vec4(s1);

    s1 = imageSize(bufSamp4);
    v *= vec4(s1);
    s1 = imageSize(bufSamp5);
    v *= vec4(s1);
    s1 = imageSize(bufSamp6);
    v *= vec4(s1);
    
    v *= texelFetch(bufSamp1, s1);
    v *= vec4(texelFetch(bufSamp2, s1));
    v *= vec4(texelFetch(bufSamp3, s1));

    return v;
}

// GL_OES_texture_cube_map_array

uniform highp writeonly imageCubeArray  CA1;
uniform highp writeonly iimageCubeArray CA2;
uniform highp writeonly uimageCubeArray CA3;
uniform highp samplerCubeArray          CA4;
uniform highp samplerCubeArrayShadow    CA5;
uniform highp isamplerCubeArray         CA6;
uniform highp usamplerCubeArray         CA7;

vec4 CAT()
{
    highp ivec3 iv = ivec3(0);
    iv += textureSize(CA4, 1);
    iv += textureSize(CA5, 1);
    iv += textureSize(CA6, 1);
    iv += textureSize(CA7, 1);

    iv += imageSize(CA1);
    iv += imageSize(CA2);
    iv += imageSize(CA3);
    
    vec4 v = vec4(iv, 1.0);
    v *= texture(CA4, vec4(0.5));
    v *= texture(CA5, vec4(0.5), 3.0);
    v *= vec4(texture(CA6, vec4(0.5)));
    v *= vec4(texture(CA7, vec4(0.5)));

    v *= textureLod(CA4, vec4(0.5), 0.24);
    v *= vec4(textureLod(CA6, vec4(0.5), 0.26));
    v *= vec4(textureLod(CA7, vec4(0.5), 0.27));

    v *= textureGrad(CA4, vec4(0.5), vec3(0.1), vec3(0.2));
    v *= vec4(textureGrad(CA6, vec4(0.5), vec3(0.1), vec3(0.2)));
    v *= vec4(textureGrad(CA7, vec4(0.5), vec3(0.1), vec3(0.2)));

    v *= textureGather(CA4, vec4(0.5));
    v *= textureGather(CA4, vec4(0.5), 2);
    v *= vec4(textureGather(CA6, vec4(0.5)));
    v *= vec4(textureGather(CA6, vec4(0.5), 1));
    v *= vec4(textureGather(CA7, vec4(0.5)));
    v *= vec4(textureGather(CA7, vec4(0.5), 0));

    v *= textureGather(CA5, vec4(0.5), 2.5);

    return v;
}

// GL_OES_texture_storage_multisample_2d_array

uniform highp sampler2DMSArray  samp2DMSA;
uniform highp isampler2DMSArray samp2DMSAi;
uniform highp usampler2DMSArray samp2DMSAu;

vec4 MSA()
{
    highp ivec3 iv = ivec3(0);
    iv += textureSize(samp2DMSA);
    iv += textureSize(samp2DMSAi);
    iv += textureSize(samp2DMSAu);
    
    vec4 v = vec4(iv, 1.0);
    v *= texelFetch(samp2DMSA, ivec3(5), 2);
    v *= vec4(texelFetch(samp2DMSAi, ivec3(5), 2));
    v *= vec4(texelFetch(samp2DMSAu, ivec3(5), 2));

    return v;
}

// GL_OES_shader_image_atomic

uniform layout(r32f)  highp  image2D im2Df;
uniform layout(r32ui) highp uimage2D im2Du;
uniform layout(r32i)  highp iimage2D im2Di;
uniform ivec2 P;

void goodImageAtom()
{
    float datf = 1.8;
    int dati = 4;
    uint datu = 7u;

    imageAtomicAdd(     im2Di, P, P.x);
    imageAtomicAdd(     im2Du, P, datu);
    imageAtomicMin(     im2Di, P, dati);
    imageAtomicMin(     im2Du, P, datu);
    imageAtomicMax(     im2Di, P, dati);
    imageAtomicMax(     im2Du, P, datu);
    imageAtomicAnd(     im2Di, P, dati);
    imageAtomicAnd(     im2Du, P, datu);
    imageAtomicOr(      im2Di, P, dati);
    imageAtomicOr(      im2Du, P, datu);
    imageAtomicXor(     im2Di, P, dati);
    imageAtomicXor(     im2Du, P, datu);
    imageAtomicExchange(im2Di, P, dati);
    imageAtomicExchange(im2Du, P, datu);
    imageAtomicExchange(im2Df, P, datf);
    imageAtomicCompSwap(im2Di, P,  3, dati);
    imageAtomicCompSwap(im2Du, P, 5u, datu);
}
