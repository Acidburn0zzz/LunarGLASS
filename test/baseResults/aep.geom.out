
Top IR:
; ModuleID = 'Glslang'

%fromVertex = type { <3 x float>, i32, i32 }
%fromVertex.0 = type { <3 x float> }
%gl_PerVertex = type { <4 x float>, float }
%gl_PerVertex.1 = type { <4 x float>, float }
%toFragment = type { <3 x float> }

@"anon@0" = global %fromVertex zeroinitializer
@fromV = global [4 x %fromVertex.0] zeroinitializer
@centr = global [4 x <3 x float>] zeroinitializer
@"anon@1" = global %gl_PerVertex zeroinitializer
@gl_in = global [4 x %gl_PerVertex.1] zeroinitializer
@gl_PrimitiveID = global i32 0
@gl_PrimitiveIDIn = global i32 0
@gl_Layer = global i32 0
@gl_InvocationID = global i32 0
@nonBlockUnsized = global [4 x <4 x float>] zeroinitializer
@toF = global %toFragment zeroinitializer

define fastcc void @main() {
entry:
  br label %mainBody

mainBody:                                         ; preds = %entry
  call void @llvm.gla.emitVertex()
  call void @llvm.gla.endPrimitive()
  %0 = load <3 x float>* getelementptr inbounds ([4 x %fromVertex.0]* @fromV, i32 0, i32 0, i32 0)
  %1 = load <3 x float>* getelementptr inbounds ([4 x <3 x float>]* @centr, i32 0, i32 1)
  %2 = fadd <3 x float> %0, %1, !gla.precision !40
  store <3 x float> %2, <3 x float>* getelementptr inbounds (%fromVertex* @"anon@0", i32 0, i32 0)
  %3 = load <4 x float>* getelementptr inbounds ([4 x %gl_PerVertex.1]* @gl_in, i32 0, i32 0, i32 0)
  store <4 x float> %3, <4 x float>* getelementptr inbounds (%gl_PerVertex* @"anon@1", i32 0, i32 0)
  %gl_PrimitiveID = load i32* @gl_PrimitiveIDIn
  store i32 %gl_PrimitiveID, i32* @gl_PrimitiveID
  store i32 2, i32* @gl_Layer
  store i32 4, i32* getelementptr inbounds (%fromVertex* @"anon@0", i32 0, i32 1)
  %4 = load i32* @gl_InvocationID
  store i32 %4, i32* getelementptr inbounds (%fromVertex* @"anon@0", i32 0, i32 2)
  call void @"pointSize2("()
  br label %stage-epilogue

stage-epilogue:                                   ; preds = %mainBody
  br label %stage-exit

stage-exit:                                       ; preds = %stage-epilogue
  ret void
}

; Function Attrs: alwaysinline
define internal fastcc void @"pointSize2("() #0 {
entry:
  %ps = alloca float
  %ps1 = load float* getelementptr inbounds ([4 x %gl_PerVertex.1]* @gl_in, i32 0, i32 3, i32 1)
  store float %ps1, float* %ps
  %0 = load float* %ps
  store float %0, float* getelementptr inbounds (%gl_PerVertex* @"anon@1", i32 0, i32 1)
  ret void
}

; Function Attrs: nounwind
declare void @llvm.gla.emitVertex() #1

; Function Attrs: nounwind
declare void @llvm.gla.endPrimitive() #1

attributes #0 = { alwaysinline }
attributes #1 = { nounwind }

!gla.invocations = !{!0}
!gla.numVertices = !{!1}
!gla.inputPrimitive = !{!2}
!gla.outputPrimitive = !{!3}
!gla.entrypoint = !{!4}
!gla.outputs = !{!5, !13, !20, !22, !24}
!gla.inputs = !{!27, !30, !32, !34, !36, !38}
!gla.noStaticUse = !{!38, !24}

!0 = metadata !{i32 4}
!1 = metadata !{i32 200}
!2 = metadata !{i32 3}
!3 = metadata !{i32 7}
!4 = metadata !{metadata !"main", i32 15}
!5 = metadata !{metadata !"", i32 16, %fromVertex* @"anon@0_typeProxy", metadata !6, metadata !7}
!6 = metadata !{i32 0, i32 0, i32 1024, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!7 = metadata !{metadata !"fromVertex", metadata !8, metadata !"color", metadata !9, metadata !"len", metadata !11, metadata !"inv", metadata !11}
!8 = metadata !{i32 0, i32 0, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!9 = metadata !{metadata !"", metadata !10}
!10 = metadata !{i32 0, i32 2, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!11 = metadata !{metadata !"", metadata !12}
!12 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!13 = metadata !{metadata !"", i32 16, %gl_PerVertex* @"anon@1_typeProxy", metadata !14, metadata !15}
!14 = metadata !{i32 0, i32 0, i32 1035, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!15 = metadata !{metadata !"gl_PerVertex", metadata !8, metadata !"gl_Position", metadata !16, metadata !"gl_PointSize", metadata !18}
!16 = metadata !{metadata !"", metadata !17}
!17 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 9, i32 -1, i32 0, i32 -1}
!18 = metadata !{metadata !"", metadata !19}
!19 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 10, i32 -1, i32 0, i32 -1}
!20 = metadata !{metadata !"gl_PrimitiveID", i32 7, i32* @gl_PrimitiveID_typeProxy, metadata !21}
!21 = metadata !{i32 0, i32 3, i32 1045, null, i32 0, i32 31, i32 -1, i32 0, i32 -1}
!22 = metadata !{metadata !"gl_Layer", i32 7, i32* @gl_Layer_typeProxy, metadata !23}
!23 = metadata !{i32 0, i32 3, i32 1047, null, i32 0, i32 32, i32 -1, i32 0, i32 -1}
!24 = metadata !{metadata !"toF", i32 16, %toFragment* @toF_typeProxy, metadata !25, metadata !26}
!25 = metadata !{i32 0, i32 0, i32 1053, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!26 = metadata !{metadata !"toFragment", metadata !8, metadata !"color", metadata !9}
!27 = metadata !{metadata !"fromV", i32 17, [4 x %fromVertex.0]* @fromV_typeProxy, metadata !28, metadata !29}
!28 = metadata !{i32 0, i32 0, i32 1027, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!29 = metadata !{metadata !"fromVertex", metadata !8, metadata !"color", metadata !9}
!30 = metadata !{metadata !"centr", i32 1, [4 x <3 x float>]* @centr_typeProxy, metadata !31}
!31 = metadata !{i32 0, i32 2, i32 1031, null, i32 512, i32 0, i32 -1, i32 0, i32 -1}
!32 = metadata !{metadata !"gl_in", i32 17, [4 x %gl_PerVertex.1]* @gl_in_typeProxy, metadata !33, metadata !15}
!33 = metadata !{i32 0, i32 0, i32 1037, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!34 = metadata !{metadata !"gl_PrimitiveIDIn", i32 1, i32* @gl_PrimitiveIDIn_typeProxy, metadata !35}
!35 = metadata !{i32 0, i32 3, i32 1046, null, i32 0, i32 31, i32 -1, i32 0, i32 -1}
!36 = metadata !{metadata !"gl_InvocationID", i32 1, i32* @gl_InvocationID_typeProxy, metadata !37}
!37 = metadata !{i32 0, i32 3, i32 1048, null, i32 0, i32 30, i32 -1, i32 0, i32 -1}
!38 = metadata !{metadata !"nonBlockUnsized", i32 1, [4 x <4 x float>]* @nonBlockUnsized_typeProxy, metadata !39}
!39 = metadata !{i32 0, i32 2, i32 1049, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!40 = metadata !{i32 2}


Bottom IR:
; ModuleID = 'Glslang'

%fromVertex = type { <3 x float>, i32, i32 }
%fromVertex.0 = type { <3 x float> }
%gl_PerVertex = type { <4 x float>, float }
%gl_PerVertex.1 = type { <4 x float>, float }
%toFragment = type { <3 x float> }

@"anon@0" = global %fromVertex zeroinitializer
@fromV = global [4 x %fromVertex.0] zeroinitializer
@centr = global [4 x <3 x float>] zeroinitializer
@"anon@1" = global %gl_PerVertex zeroinitializer
@gl_in = global [4 x %gl_PerVertex.1] zeroinitializer
@gl_PrimitiveID = global i32 0
@gl_PrimitiveIDIn = global i32 0
@gl_Layer = global i32 0
@gl_InvocationID = global i32 0
@nonBlockUnsized = global [4 x <4 x float>] zeroinitializer
@toF = global %toFragment zeroinitializer

define fastcc void @main() {
entry:
  call void @llvm.gla.emitVertex()
  call void @llvm.gla.endPrimitive()
  %gla_constGEP = getelementptr [4 x %fromVertex.0]* @fromV, i32 0, i32 0, i32 0
  %0 = load <3 x float>* %gla_constGEP, align 16
  %gla_constGEP1 = getelementptr [4 x <3 x float>]* @centr, i32 0, i32 1
  %1 = load <3 x float>* %gla_constGEP1, align 16
  %2 = fadd <3 x float> %0, %1, !gla.precision !40
  %gla_constGEP2 = getelementptr %fromVertex* @"anon@0", i32 0, i32 0
  store <3 x float> %2, <3 x float>* %gla_constGEP2, align 16
  %gla_constGEP3 = getelementptr [4 x %gl_PerVertex.1]* @gl_in, i32 0, i32 0, i32 0
  %3 = load <4 x float>* %gla_constGEP3, align 16
  %gla_constGEP4 = getelementptr %gl_PerVertex* @"anon@1", i32 0, i32 0
  store <4 x float> %3, <4 x float>* %gla_constGEP4, align 16
  %gl_PrimitiveID = load i32* @gl_PrimitiveIDIn, align 4
  store i32 %gl_PrimitiveID, i32* @gl_PrimitiveID, align 4
  store i32 2, i32* @gl_Layer, align 4
  %gla_constGEP5 = getelementptr %fromVertex* @"anon@0", i32 0, i32 1
  store i32 4, i32* %gla_constGEP5, align 16
  %4 = load i32* @gl_InvocationID, align 4
  %gla_constGEP6 = getelementptr %fromVertex* @"anon@0", i32 0, i32 2
  store i32 %4, i32* %gla_constGEP6, align 4
  %gla_constGEP7 = getelementptr [4 x %gl_PerVertex.1]* @gl_in, i32 0, i32 3, i32 1
  %ps1.i = load float* %gla_constGEP7, align 16
  %gla_constGEP8 = getelementptr %gl_PerVertex* @"anon@1", i32 0, i32 1
  store float %ps1.i, float* %gla_constGEP8, align 16
  br label %stage-epilogue

stage-epilogue:                                   ; preds = %entry
  br label %stage-exit

stage-exit:                                       ; preds = %stage-epilogue
  ret void
}

; Function Attrs: nounwind
declare void @llvm.gla.emitVertex() #0

; Function Attrs: nounwind
declare void @llvm.gla.endPrimitive() #0

attributes #0 = { nounwind }

!gla.invocations = !{!0}
!gla.numVertices = !{!1}
!gla.inputPrimitive = !{!2}
!gla.outputPrimitive = !{!3}
!gla.entrypoint = !{!4}
!gla.outputs = !{!5, !13, !20, !22, !24}
!gla.inputs = !{!27, !30, !32, !34, !36, !38}
!gla.noStaticUse = !{!38, !24}

!0 = metadata !{i32 4}
!1 = metadata !{i32 200}
!2 = metadata !{i32 3}
!3 = metadata !{i32 7}
!4 = metadata !{metadata !"main", i32 15}
!5 = metadata !{metadata !"", i32 16, %fromVertex* @"anon@0_typeProxy", metadata !6, metadata !7}
!6 = metadata !{i32 0, i32 0, i32 1024, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!7 = metadata !{metadata !"fromVertex", metadata !8, metadata !"color", metadata !9, metadata !"len", metadata !11, metadata !"inv", metadata !11}
!8 = metadata !{i32 0, i32 0, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!9 = metadata !{metadata !"", metadata !10}
!10 = metadata !{i32 0, i32 2, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!11 = metadata !{metadata !"", metadata !12}
!12 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!13 = metadata !{metadata !"", i32 16, %gl_PerVertex* @"anon@1_typeProxy", metadata !14, metadata !15}
!14 = metadata !{i32 0, i32 0, i32 1035, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!15 = metadata !{metadata !"gl_PerVertex", metadata !8, metadata !"gl_Position", metadata !16, metadata !"gl_PointSize", metadata !18}
!16 = metadata !{metadata !"", metadata !17}
!17 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 9, i32 -1, i32 0, i32 -1}
!18 = metadata !{metadata !"", metadata !19}
!19 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 10, i32 -1, i32 0, i32 -1}
!20 = metadata !{metadata !"gl_PrimitiveID", i32 7, i32* @gl_PrimitiveID_typeProxy, metadata !21}
!21 = metadata !{i32 0, i32 3, i32 1045, null, i32 0, i32 31, i32 -1, i32 0, i32 -1}
!22 = metadata !{metadata !"gl_Layer", i32 7, i32* @gl_Layer_typeProxy, metadata !23}
!23 = metadata !{i32 0, i32 3, i32 1047, null, i32 0, i32 32, i32 -1, i32 0, i32 -1}
!24 = metadata !{metadata !"toF", i32 16, %toFragment* @toF_typeProxy, metadata !25, metadata !26}
!25 = metadata !{i32 0, i32 0, i32 1053, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!26 = metadata !{metadata !"toFragment", metadata !8, metadata !"color", metadata !9}
!27 = metadata !{metadata !"fromV", i32 17, [4 x %fromVertex.0]* @fromV_typeProxy, metadata !28, metadata !29}
!28 = metadata !{i32 0, i32 0, i32 1027, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!29 = metadata !{metadata !"fromVertex", metadata !8, metadata !"color", metadata !9}
!30 = metadata !{metadata !"centr", i32 1, [4 x <3 x float>]* @centr_typeProxy, metadata !31}
!31 = metadata !{i32 0, i32 2, i32 1031, null, i32 512, i32 0, i32 -1, i32 0, i32 -1}
!32 = metadata !{metadata !"gl_in", i32 17, [4 x %gl_PerVertex.1]* @gl_in_typeProxy, metadata !33, metadata !15}
!33 = metadata !{i32 0, i32 0, i32 1037, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!34 = metadata !{metadata !"gl_PrimitiveIDIn", i32 1, i32* @gl_PrimitiveIDIn_typeProxy, metadata !35}
!35 = metadata !{i32 0, i32 3, i32 1046, null, i32 0, i32 31, i32 -1, i32 0, i32 -1}
!36 = metadata !{metadata !"gl_InvocationID", i32 1, i32* @gl_InvocationID_typeProxy, metadata !37}
!37 = metadata !{i32 0, i32 3, i32 1048, null, i32 0, i32 30, i32 -1, i32 0, i32 -1}
!38 = metadata !{metadata !"nonBlockUnsized", i32 1, [4 x <4 x float>]* @nonBlockUnsized_typeProxy, metadata !39}
!39 = metadata !{i32 0, i32 2, i32 1049, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!40 = metadata !{i32 2}
#version 310 es
// LunarGOO output
#extension GL_ANDROID_extension_pack_es31a : enable
#extension GL_EXT_geometry_shader : enable
#extension GL_EXT_gpu_shader5 : enable
#extension GL_EXT_primitive_bounding_box : enable
#extension GL_EXT_shader_io_blocks : enable
#extension GL_EXT_tessellation_shader : enable
#extension GL_EXT_texture_buffer : enable
#extension GL_EXT_texture_cube_map_array : enable
#extension GL_KHR_blend_equation_advanced : enable
#extension GL_OES_geometry_point_size : enable
#extension GL_OES_sample_variables : enable
#extension GL_OES_shader_image_atomic : enable
#extension GL_OES_shader_multisample_interpolation : enable
#extension GL_OES_texture_storage_multisample_2d_array : enable
layout(lines_adjacency) in;
layout(invocations = 4) in;
layout(triangle_strip) out;
layout(max_vertices = 200) out;
in fromVertex {
	mediump vec3 color;
} fromV[4];
in mediump vec3 centr[4];
in mediump vec4 nonBlockUnsized[4];
out fromVertex {
	mediump vec3 color;
	highp int len;
	highp int inv;
} ;
out toFragment {
	mediump vec3 color;
} toF;
const int C_2 = 2;
const int C_4 = 4;

void main()
{
	EmitVertex();
	EndPrimitive();
	mediump vec3 H_g00yqf1 = centr[1] + fromV[0].color;
	color = H_g00yqf1;
	gl_Position = gl_in[0].gl_Position;
	gl_PrimitiveID = gl_PrimitiveIDIn;
	gl_Layer = C_2;
	len = C_4;
	inv = gl_InvocationID;
	gl_PointSize = gl_in[3].gl_PointSize;
	
}

#version 310 es
// LunarGOO output
#extension GL_ANDROID_extension_pack_es31a : enable
#extension GL_EXT_geometry_shader : enable
#extension GL_EXT_gpu_shader5 : enable
#extension GL_EXT_primitive_bounding_box : enable
#extension GL_EXT_shader_io_blocks : enable
#extension GL_EXT_tessellation_shader : enable
#extension GL_EXT_texture_buffer : enable
#extension GL_EXT_texture_cube_map_array : enable
#extension GL_KHR_blend_equation_advanced : enable
#extension GL_OES_geometry_point_size : enable
#extension GL_OES_sample_variables : enable
#extension GL_OES_shader_image_atomic : enable
#extension GL_OES_shader_multisample_interpolation : enable
#extension GL_OES_texture_storage_multisample_2d_array : enable
layout(lines_adjacency) in;
layout(invocations = 4) in;
layout(triangle_strip) out;
layout(max_vertices = 200) out;
in mediump vec3 centr[4];
in fromVertex {
	mediump vec3 color;
} fromV[4];
in mediump vec4 nonBlockUnsized[4];
out fromVertex {
	mediump vec3 color;
	highp int len;
	highp int inv;
} ;
out toFragment {
	mediump vec3 color;
} toF;
const int C_2 = 2;
const int C_4 = 4;

void main()
{
	EmitVertex();
	EndPrimitive();
	mediump vec3 H_g00yqf = centr[1] + fromV[0].color;
	color = H_g00yqf;
	gl_Position = gl_in[0].gl_Position;
	gl_PrimitiveID = gl_PrimitiveIDIn;
	gl_Layer = C_2;
	len = C_4;
	inv = gl_InvocationID;
	gl_PointSize = gl_in[3].gl_PointSize;
	
}

